qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_presenter.py

```diff
@@ -49,39 +43,81 @@ def notify_data_changed(self):
+        if last_row >= 0:
+            self._view.get_pairing_table.removeRow(last_row)
+
+    def clear(self):
+        # Go backwards to preserve indices
+        for row in reversed(range(self.num_rows())):
+            self._view.get_pairing_table.removeRow(row)
+
+    def get_selected_row_indices(self):
+        return list(set(index.row() for index in self._view.get_pairing_table.selectedIndexes()))
+
+    def get_selected_pair_names_and_indexes(self):
+        indexes = self.get_selected_row_indices()
+        return [[str(self._view.get_pairing_table.item(i, 0).text()), i] for i in indexes]
 
     def handle_guess_alpha_clicked(self, row):
-        table_row = self._view.get_table_contents()[row]
+        table_row = self.get_table_contents()[row]
         pair_name = table_row[0]
         group1 = table_row[2]
         group2 = table_row[3]
         self.guessAlphaNotifier.notify_subscribers([pair_name, group1, group2])
 
     def handle_data_change(self, row, col):
-        table = self._view.get_table_contents()
+        table = self.get_table_contents()
         changed_item = self._view.get_table_item(row, col)
-        changed_item_text = self._view.get_table_item_text(row, col)
-        pair_name = self._view.get_table_item_text(row, 0)
+        changed_item_text = self.get_table_item_text(row, col)
+        pair_name = self.get_table_item_text(row, 0)
         update_model = True
-        if pair_columns[col] == "pair_name" and not self.validate_pair_name(changed_item_text):
+        if get_pair_columns()[col] == "pair_name" and not self.validate_pair_name(changed_item_text):
             update_model = False
-        if pair_columns[col] == "group_1":
-            if changed_item_text == self._view.get_table_item_text(row, pair_columns.index("group_2")):
-                table[row][pair_columns.index("group_2")] = self._model.pairs[row].forward_group
-        if pair_columns[col] == "group_2":
-            if changed_item_text == self._view.get_table_item_text(row, pair_columns.index("group_1")):
-                table[row][pair_columns.index("group_1")] = self._model.pairs[row].backward_group
-        if pair_columns[col] == "alpha":
+        if get_pair_columns()[col] == "group_1":
+            if changed_item_text == self.get_table_item_text(row, get_pair_columns().index("group_2")):
+                table[row][get_pair_columns().index("group_2")] = self._model.pairs[row].forward_group
+        if get_pair_columns()[col] == "group_2":
+            if changed_item_text == self.get_table_item_text(row, get_pair_columns().index("group_1")):
+                table[row][get_pair_columns().index("group_1")] = self._model.pairs[row].backward_group
+        if get_pair_columns()[col] == "alpha":
             if not self.validate_alpha(changed_item_text):
                 update_model = False
             else:
                 rounded_item = round_value(changed_item_text, self._model._context.group_pair_context.alpha_precision)
```

Comment:
In the presenter class, a private member of the model : `self.model._context` is accesed in a few places. Can you add the necessary getter functions on this model so that the context is not accesed directly?